### PR TITLE
Add methods to User model to change/test status [WIP]

### PR DESCRIPTION
**Pull request by seafr**
* [x] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `rake test`
* [x] code is in uniquely-named feature branch and has no merge conflicts
* [x] PR is descriptively titled

Thanks!

### CODE REVIEW

**Comment by seafr** [comment]
There is one method missing to complete the trio of whether the user is banned, moderated or none of these. I am not sure 'active' would do and I need help determining an appropriate name for this. Please suggest some names. Thanks.

**Comment by SidharthBansal** [comment]
@seafr how about 'normal' ?

**Comment by seafr** [comment]
@SidharthBansal I've thought about this. And about 'active' too. The "problem" is when you realise what their opposites are.

File: app/models/user.rb [hunk 1]
```diff
@@ -159,6 +159,22 @@ def can_moderate?
     admin? || moderator?
   end
 
+  # User status
+  # 0: banned
+  # 5: moderated
+  def mark_as!(new_status)
+    self.status = 0 if new_status == :banned
+    self.status = 5 if new_status == :moderated
```
Comment (siaw23) [inline app/models/user.rb]: let's use ternary operator here for simplicity if this is only going to be about 2 states.

Comment (shauncarland) [reply app/models/user.rb]: Did you consider using an enum instead of adding this definition?
https://bendyworks.com/blog/rails-enum-sharp-knife

**Comment by shauncarland** [comment]
Did you consider using an enum instead of adding this definition?
https://bendyworks.com/blog/rails-enum-sharp-knife

Adding something like `enum status {banned: 0, moderated: 5}` to the class model.  If you do that, you'll get access to `user.banned?`, `user.moderated?` methods automatically thanks to Rails

File: app/models/user.rb (hunk 1 above)
Comment (seafr) [reply app/models/user.rb]: In fact, 1 more state is missing. Check my first comment above (and please help find a good name :) ). 
Thank you for review.

File: test/unit/user_test.rb [hunk 2]
```diff
@@ -184,6 +185,16 @@ class UserTest < ActiveSupport::TestCase
     assert_not basic_user.moderator?
     assert_not basic_user.can_moderate?
   end
+
+  test 'user statuses' do
```
Comment (Gauravano) [inline test/unit/user_test.rb]: if possible, keep it status only, totally your call

File: app/models/user.rb [hunk 3]
```diff
@@ -159,6 +159,22 @@ def can_moderate?
     admin? || moderator?
   end
 
+  # User status
+  # 0: banned
+  # 5: moderated
+  def mark_as!(new_status)
```
Comment (Gauravano) [inline app/models/user.rb]: hi @seafr,  changes look good, I would agree with 'normal' as status as for comments we use 'normal' only

**Comment by jywarren** [comment]
Thanks for the careful and thorough teamwork here! It's a tough naming choice and I was thinking of `basic` for a bit, but I think a moderated user is `basic` too -- the word we need is to contrast with a moderated status. So, I agree that `active` seems best here. 😄 👍

**Comment by seafr** [comment]
Sorry for delay guys. 

So, I'm going with @jywarren's choice of `active`. Thank you all for valuable input.

**Comment by seafr** [comment]
@jywarren I'm stuck here. Can you please take a look?

File: app/models/user.rb [hunk 4]
```diff
@@ -41,6 +41,12 @@ class User < ActiveRecord::Base
   before_save :set_token
   after_destroy :destroy_drupal_user
 
+  # User status
+  # 0: banned
+  # 1: active
+  # 5: moderated
+  enum status: {banned: 0, active: 1, moderated: 5}
```
Comment (jywarren) [inline app/models/user.rb]: Wow, i have never used `enum` like this. Can you link me to some documentation to follow how it works?

**Comment by jywarren** [comment]
Ah, it looks like the tests are finding a `logged out` error instead of a `The user 'rob' has been placed in moderation...` message! Can you figure out why?

```
 FAIL["test_users_are_logged_out_and_alerted_when_moderated,_and_notes_are_not_accessible", Minitest::Result, 8.989677048000033]
 test_users_are_logged_out_and_alerted_when_moderated,_and_notes_are_not_accessible#Minitest::Result (8.99s)
        --- expected
        +++ actual
        @@ -1 +1 @@
        -"The user 'rob' has been placed in moderation; please see <a href='https://www.example.com/wiki/moderators'>our moderation policy</a> and contact <a href='mailto:moderators@www.example.com'>moderators@www.example.com</a> if you believe this is in error."
        +"You must be logged in to access this page"
        test/integration/moderate_and_ban_test.rb:68:in `block in <class:ModerateAndBanTest>'
```

**Comment by jywarren** [comment]
Hi, just checking if you've gotten stuck on this at all, or if I could help in any way? Thanks!

**Comment by gauravano** [comment]
Hi @jywarren do you think this PR is required as we have deprecated drupal user? Let me know,if there is still some potential in this PR for optimizing codebase, I may be able to help. Thanks!

**Comment by jywarren** [comment]
This enum approach to user status does look pretty useful and efficient. Perhaps we can try to implement just that part?

**Comment by gauravano** [comment]
Hi everyone, thanks for working and contributing to this PR. @siaw  did some cleanup in https://github.com/publiclab/plots2/pull/4894/files which cover the same goal as this PR, so I think, we can close this one. Feel free to open this PR again(actually, prefer to open new PR as we have many changes since then), if you think, there's still more for improvement.  

Thanks all! 

cc @jywarren

**Pull request closed without merging**